mlserver/batching/requests.py

```diff
@@ -0,0 +1,145 @@
+from operator import mul
+from collections import defaultdict
+from typing import Any, Dict, Iterable, List, Tuple, Union
+from functools import reduce
+
+from ..types import (
+    InferenceRequest,
+    InferenceResponse,
+    RequestInput,
+    ResponseOutput,
+)
+
+
+def _get_data(payload: Union[RequestInput, ResponseOutput]):
+    return getattr(payload.data, "__root__", payload.data)
+
+
+def _infer_elem_size(response_output: ResponseOutput) -> int:
+    batch_axis = _get_batch_axis(response_output)
+    shape_without_batch = response_output.shape.copy()
+    shape_without_batch.pop(batch_axis)
+    return reduce(mul, shape_without_batch, 1)
+
+
+def _get_batch_size(request_input: RequestInput) -> int:
+    batch_dim = _get_batch_axis(request_input)
+    return request_input.shape[batch_dim]
+
+
+def _get_batch_axis(
+    request_input: Union[RequestInput, ResponseOutput]  # type: ignore
+) -> int:
+    # TODO: Support other batch dimensions
+    return 0
+
+
+class BatchedRequests:
+    def __init__(self, inference_requests: List[InferenceRequest] = []):
+        self._inference_requests = inference_requests
+        self._prediction_ids: List[str] = [  # type: ignore
+            req.id for req in self._inference_requests  # type: ignore
+        ]
+        self._minibatch_sizes: List[int] = []
+        self.merged_request = self._merge_requests()
+
+    def _merge_requests(self) -> InferenceRequest:
+        inputs_index: Dict[str, List[RequestInput]] = defaultdict(list)
+
+        for inference_request in self._inference_requests:
+            # TODO: What should happen if UID is empty?
+            self._minibatch_sizes = []
+            for request_input in inference_request.inputs:
+                inputs_index[request_input.name].append(request_input)
+
+        inputs = [
+            self._merge_request_inputs(request_inputs)
+            for request_inputs in inputs_index.values()
+        ]
+
+        # TODO: Add outputs
+        # TODO: Should we add a 'fake' request ID?
+        return InferenceRequest(inputs=inputs)
+
+    def _merge_request_inputs(self, request_inputs: List[RequestInput]) -> RequestInput:
+        batch_size, data = self._merge_data(request_inputs)
+
+        # TODO: What should we do if list is empty?
+        sampled = request_inputs[0]
+
+        batch_axis = _get_batch_axis(sampled)
+        shape = sampled.shape.copy()
+        shape.pop(batch_axis)
+        shape.insert(batch_axis, batch_size)
+
+        return RequestInput(
+            name=sampled.name, datatype=sampled.datatype, shape=shape, data=data
+        )
+
+    def _merge_data(self, request_inputs: List[RequestInput]) -> Tuple[int, Any]:
+        # We assume all inputs will have the same minibatch size
+        self._minibatch_sizes = []
```

Comment:
is that when the inputs are already coming batched from the client?

Reply:
Yes, exactly. I'll add some comments to clarify what do we mean by minibatches.